ouroboros-consensus/src/Ouroboros/Consensus/Node.hs

```diff
@@ -143,6 +165,115 @@ run tracers chainDbTracer rna dbPath pInfo
+-- invalid and will be deleted. This would throw away a perfectly good chain,
+-- possibly consisting of gigabytes of data that will have to be synched
+-- again.
+--
+-- To protect us from unwanted deletion of valid files, we first check whether
+-- we have been given the path to the right database folder. We do this by
+-- reading the 'ProtocolMagicId' of the net from a file stored in the root of
+-- the database folder. This file's name is defined in 'protocolMagicIdFile'.
+--
+-- * If the 'ProtocolMagicId' from the file matches that of the net, we have
+--   the right database folder.
+-- * If not, we are opening the wrong database folder and abort by throwing a
+--   'ChainDbCheckError'.
+-- * If there is no such file and the folder is empty, we create it and store
+--   the net's 'ProtocolMagicId' in it.
+-- * If there is no such file, but the folder is not empty, we throw a
+--   'ChainDbCheckError', because we have likely been given the wrong path,
+--   maybe to a folder containing user or system files. This includes the case
+--   that the 'protocolMagicIdFile' has been deleted.
+-- * If there is such a 'protocolMagicIdFile', but it could not be read or its
+--   contents could not be parsed, we also throw a 'ChainDbCheckError'.
+--
+-- Note that an 'FsError' can also be thrown.
+checkProtocolMagicId
+  :: forall m h. MonadThrow m
+  => HasFS m h
+  -> MountPoint
+     -- ^ Database directory. Should be the mount point of the @HasFS@. Used
+     -- in error messages.
+  -> ProtocolMagicId
+  -> m (Either ChainDbCheckError ())
+checkProtocolMagicId hasFS mountPoint protocolMagicId = runExceptT $ do
+    lift $ createDirectoryIfMissing hasFS True root
+    fileExists <- lift $ doesFileExist hasFS pFile
+    if fileExists then do
+      actualProtocolMagicId <- readProtocolMagicIdFile
+      when (actualProtocolMagicId /= protocolMagicId) $
+        throwError $ ProtocolMagicIdMismatch
+          fullPath
+          actualProtocolMagicId
+          protocolMagicId
+    else do
+      isEmpty <- lift $ Set.null <$> listDirectory hasFS root
+      if isEmpty then
+        createProtocolMagicIdFile
+      else
+        throwError $ NoProtocolMagicIdAndNotEmpty fullPath
+  where
+    root     = mkFsPath []
+    pFile    = fsPathFromList [protocolMagicIdFile]
+    fullPath = fsToFilePath mountPoint pFile
+
+    readProtocolMagicIdFile :: ExceptT ChainDbCheckError m ProtocolMagicId
+    readProtocolMagicIdFile = ExceptT $
+      withFile hasFS pFile ReadMode $ \h -> do
+        res <- CBOR.deserialiseFromBytes fromCBOR <$> hGetAll hasFS h
+        case res of
+          Right (bl, a) | Lazy.null bl -> return $ Right a
+          _ -> return $ Left $ CorruptProtocolMagicId fullPath
+
+    createProtocolMagicIdFile :: ExceptT ChainDbCheckError m ()
+    createProtocolMagicIdFile = lift $
+      withFile hasFS pFile (AppendMode MustBeNew) $ \h ->
+        void $ hPutAll hasFS h $
+          CBOR.toLazyByteString (toCBOR protocolMagicId)
```

Comment:
Alternatively, we use `show` + `readMaybe` on the underlying `Word32` to get a readable value, which would make debugging easier.

Reply:
Yeah, let's do this. There is precedent for printing `ProtocolMagic` as a decimal number.